unit-tests/src/main/scala/java/lang/DoubleSuite.scala

```diff
@@ -1,6 +1,191 @@
+    val pinf2 = scala.Double.MaxValue + scala.Double.MaxValue
+    assert(pinf1 equals pinf2)
+
+    val bpinf1: java.lang.Double = pinf1
+    val bpinf2: java.lang.Double = pinf2
+    assert(bpinf1 equals bpinf2)
+
+    val ninf1 = scala.Double.NegativeInfinity
+    val ninf2 = scala.Double.MinValue + scala.Double.MinValue
+    assert(ninf1 equals ninf2)
+
+    val bninf1: java.lang.Double = ninf1
+    val bninf2: java.lang.Double = ninf2
+    assert(bninf1 equals bninf2)
+
+    assert(Double.NaN equals Double.NaN)
+
+    val x = Double.NaN
+    val y = longBitsToDouble(doubleToRawLongBits(x) | 1)
+    assert(x equals y)
+
+    val z = longBitsToDouble(doubleToLongBits(x) | 1)
+    assert(x equals z)
+  }
+
+  test("==") {
+    val pzero = +0.0
+    val nzero = -0.0
+    assert(pzero == pzero)
+    assert(nzero == nzero)
+    assert(pzero == nzero)
+    val szero = 1.0 - 1.0
+    assert(pzero == szero)
+
+    val bpzero: java.lang.Double = pzero
+    val bnzero: java.lang.Double = nzero
+    assertNot(bpzero == bnzero)
+    val bszero: java.lang.Double = szero
+    assert(bpzero == bszero)
+
+    val num1 = 123.45
+    val num2 = 123.45
+    assert(num1 == num2)
+
+    val bnum1: java.lang.Double = num1
+    val bnum2: java.lang.Double = num2
+    assert(bnum1 == bnum2)
+
+    val pmax1 = scala.Double.MaxValue
+    val pmax2 = scala.Double.MaxValue
+    assert(pmax1 == pmax2)
+    val pmax3 = scala.Double.MaxValue + 1
+    assert(pmax1 == pmax3)
+
+    val bpmax1: java.lang.Double = scala.Double.MaxValue
+    val bpmax2: java.lang.Double = scala.Double.MaxValue
+    assert(bpmax1 == bpmax2)
+    val bpmax3: java.lang.Double = scala.Double.MaxValue + 1
+    assert(bpmax1 == bpmax3)
+
+    val pmin1 = scala.Double.MinValue
+    val pmin2 = scala.Double.MinValue
+    assert(pmin1 == pmin2)
+    val pmin3 = scala.Double.MinValue + 1
+    assert(pmin1 == pmin3)
+
+    val bpmin1: java.lang.Double = scala.Double.MinValue
+    val bpmin2: java.lang.Double = scala.Double.MinValue
+    assert(bpmin1 == bpmin2)
+    val bpmin3: java.lang.Double = scala.Double.MinValue + 1
+    assert(bpmin1 == bpmin3)
+
+    val pinf1 = scala.Double.PositiveInfinity
+    val pinf2 = scala.Double.MaxValue + scala.Double.MaxValue
+    assert(pinf1 == pinf2)
+
+    val bpinf1: java.lang.Double = pinf1
+    val bpinf2: java.lang.Double = pinf2
+    assert(bpinf1 == bpinf2)
+
+    val ninf1 = scala.Double.NegativeInfinity
+    val ninf2 = scala.Double.MinValue + scala.Double.MinValue
+    assert(ninf1 == ninf2)
+
+    val bninf1: java.lang.Double = ninf1
+    val bninf2: java.lang.Double = ninf2
+    assert(bninf1 == bninf2)
+
+    assertNot(Double.NaN == Double.NaN)
+
+    val x = Double.NaN
+    val y = longBitsToDouble(doubleToRawLongBits(x) | 1)
+    // assertNot(x == y) // FIXME this works on the JVM, but fails on native.
+
+    val z = longBitsToDouble(doubleToLongBits(x) | 1)
+    // assertNot(x == z) // FIXME this works on the JVM, but fails on native.
```

Comment:
One reason why this might happen is that we use `fcmp ueq` which returns `true` if either operand is unordered (NaN). Maybe `fcmp oeq` captures the semantics of the JVM's == for floating point numbers more precisely?